core/drivers/versal_net_nvm.c

```diff
@@ -0,0 +1,1067 @@
+{
+	struct versal_nvm_write_req req __aligned_efuse = {
+		.efuse_id = BBRAM_WRITE_USER_DATA,
+		.bbram.user_data = data,
+	};
+
+	if (versal_nvm_write(&req))
+		return TEE_ERROR_GENERIC;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result versal_bbram_read_user_data(uint32_t *data)
+{
+	struct versal_nvm_read_req req = {
+		.efuse_id = BBRAM_READ_USER_DATA,
+	};
+
+	if (versal_alloc_read_buffer(&req))
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	if (versal_nvm_read(&req)) {
+		versal_free_read_buffer(&req);
+		return TEE_ERROR_GENERIC;
+	}
+
+	memcpy(data, versal_get_read_buffer(&req), sizeof(*data));
+	versal_free_read_buffer(&req);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result versal_bbram_lock_write_user_data(void)
+{
+	struct versal_nvm_write_req req __aligned_efuse  = {
+		.efuse_id = BBRAM_LOCK_WRITE_USER_DATA,
+	};
+
+	if (versal_nvm_write(&req))
+		return TEE_ERROR_GENERIC;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result versal_efuse_read_cache(uint16_t off, uint16_t num,
+					  uint32_t *buf, size_t len)
+{
+	struct versal_ipi_cmd cmd = { };
+	struct versal_mbox_mem p = { };
+	TEE_Result ret = TEE_SUCCESS;
+	uint32_t a = 0;
+	uint32_t b = 0;
+
+	if (!buf)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (len < num * NVM_WORD_LEN)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	ret = versal_mbox_alloc(num * NVM_WORD_LEN, NULL, &p);
+	if (ret)
+		return ret;
+
+	reg_pair_from_64(virt_to_phys(p.buf), &b, &a);
+
+	cmd.data[0] = NVM_API_ID(EFUSE_READ_CACHE);
+	cmd.data[1] = (num << 16) | off;
+	cmd.data[2] = a;
+	cmd.data[3] = b;
+
+	cmd.ibuf[0].mem = p;
+
+	ret = versal_mbox_notify_pmc(&cmd, NULL, NULL);
+	if (ret)
+		EMSG("Mailbox error");
+	else
+		memcpy(buf, p.buf, num * NVM_WORD_LEN);
+
+	versal_mbox_free(&p);
+	return ret;
+}
+
+#define EFUSE_CACHE_DNA_OFFSET				0x20
+#define EFUSE_CACHE_BOOT_ENV_CTRL_OFFSET		0x94
+#define EFUSE_CACHE_MISC_CTRL_OFFSET			0xA0
+#define EFUSE_CACHE_PUF_ECC_CTRL_OFFSET			0xA4
+#define EFUSE_CACHE_PUF_CHASH_OFFSET			0xA8
+#define EFUSE_CACHE_SEC_CTRL_OFFSET			0xAC
+#define EFUSE_CACHE_REVOCATION_ID0_OFFSET		0xB0
+#define EFUSE_CACHE_SEC_MISC0_OFFSET			0xE4
+#define EFUSE_CACHE_SEC_MISC1_OFFSET			0xE8
+#define EFUSE_CACHE_PPK0_OFFSET				0x100
+#define EFUSE_CACHE_PPK1_OFFSET				0x120
+#define EFUSE_CACHE_PPK2_OFFSET				0x140
+#define EFUSE_CACHE_OFFCHIP_REVOKE_ID0_OFFSET		0x160
+#define EFUSE_CACHE_METAHEADER_IV_RANGE0_OFFSET		0x180
+#define EFUSE_CACHE_BLACK_IV0_OFFSET			0x1D0
+#define EFUSE_CACHE_PLM_IV_RANGE0_OFFSET		0x1DC
+#define EFUSE_CACHE_DATA_PARTITION_IV_RANGE0_OFFSET	0x1E8
+#define EFUSE_CACHE_USER0_OFFSET			0x240
+#define EFUSE_CACHE_PUF_SYN0_OFFSET			0x300
+
+TEE_Result versal_efuse_read_dna(uint32_t *buf, size_t len)
+{
+	if (len < EFUSE_DNA_LEN)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return versal_efuse_read_cache(EFUSE_CACHE_DNA_OFFSET,
+				       EFUSE_DNA_LEN / NVM_WORD_LEN, buf, len);
+}
+
+TEE_Result versal_efuse_read_user_data(uint32_t *buf, size_t len,
+				       uint32_t first, size_t num)
+{
+	uint16_t offset;
```

Comment:
> Initialization value missing. Many other occurrences in this source file.

Hi @etienne-lms as these variables are not volatile is initialization necessary as the compiler will optimize the init out since offset is populated just below at L397. Should these be volatile as well to ensure initialization?

Reply:
Fixed in fixup commit 8f59535